src/Language/Haskell/Liquid/Types/PrettyPrint.hs

```diff
@@ -436,8 +450,102 @@ instance (PPrint r, F.Reftable r) => PPrint (UReft r) where
 
--- | Similar in spirit to 'reportErrors' from the GHC API, but it uses our pretty-printer
--- and shim functions under the hood.
-reportErrors :: (Show e, F.PPrint e) => F.Tidy -> [TError e] -> Ghc.TcRn ()
-reportErrors k errs =
-  forM errs (\err -> mkLongErrAt (pos err) (ppError k empty err) mempty) >>= Ghc.reportErrors
+-- | Similar in spirit to 'reportErrors' from the GHC API, but it uses our
+-- pretty-printer and shim functions under the hood. Also filters the errors
+-- according to the given `Filter` list.
+--
+-- @filterReportErrors failure continue filters k@ will call @failure@ if there
+-- are unexpected errors, or will call @continue@ otherwise.
+--
+-- An error is expected if there is any filter that matches it.
+filterReportErrors :: forall e' a. (Show e', F.PPrint e') => FilePath -> Ghc.TcRn a -> Ghc.TcRn a -> [Filter] -> F.Tidy -> [TError e'] -> Ghc.TcRn a
+filterReportErrors path failure continue filters k =
+  filterReportErrorsWith
+    FilterReportErrorsArgs { msgReporter = Ghc.reportErrors
+                           , filterReporter = defaultFilterReporter path
+                           , failure = failure
+                           , continue = continue
+                           , pprinter = \err -> mkLongErrAt (pos err) (ppError k empty err) mempty
+                           , filterMapper = reduceFilters renderer filters
+                           , filters = filters
+                           }
+  where
+    renderer :: TError e' -> String
+    renderer = render . ppError k empty
+
+-- | Retrieve the `Filter`s from the Config.
+getFilters :: Config -> [Filter]
+getFilters cfg = anyFilter <> stringFilters
+  where
+    anyFilter = [AnyFilter | expectAnyError cfg]
+    stringFilters = StringFilter <$> expectErrorContaining cfg
+
+-- | Return either the error as a singleton or the list of @filters@ that
+-- matched the @err@, given a @renderer@ for the @err@ and some @filters@
+reduceFilters :: forall e. (e -> String) -> [Filter] -> e -> Either [e] [Filter]
+reduceFilters renderer fs err =
+  if null matchingFilters
+  then Left [err]
+  else Right matchingFilters
+  where
+    matchingFilters :: [Filter]
+    matchingFilters = filter (filterDoesMatchErr err) fs
+
+    filterDoesMatchErr :: e -> Filter -> Bool
+    filterDoesMatchErr _ AnyFilter = True
+    filterDoesMatchErr e (StringFilter filter) = filter `L.isInfixOf` renderer e
+
+-- | Used in `filterReportErrorsWith'`
+data FilterReportErrorsArgs m filter msg e a =
+  FilterReportErrorsArgs
+  { msgReporter :: [msg] -> m ()
+  , filterReporter :: [filter] -> m ()
+  , failure :: m a
+  , continue :: m a
+  , pprinter :: e -> m msg
+  , filterMapper :: e -> Either [e] [filter]
+  , filters :: [filter]
+  }
+
+-- | Abstract filtered error reporter
+filterReportErrorsWith :: (Monad m, Ord filter) => FilterReportErrorsArgs m filter msg e a -> [e] -> m a
+filterReportErrorsWith FilterReportErrorsArgs {..} errs =
```

Comment:
Please, explain what the function is supposed to do. The description is insufficient to understand the properties that are being tested in the testsuite.